internal/parser/parser.go

```diff
@@ -0,0 +1,284 @@
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns/filter"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns/formatter/textcolumns"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/snapshotcombiner"
+)
+
+type ErrorCallback func(severity logger.Level, message string)
+
+// Parser is the (untyped) interface used for parser
+type Parser interface {
+	// GetTextColumnsFormatter returns the default formatter for this columns instance
+	GetTextColumnsFormatter(options ...textcolumns.Option) TextColumnsFormatter
+
+	// GetColumnNamesAndDescription returns a map of column names to their respective descriptions
+	GetColumnNamesAndDescription() map[string]string
+
+	// GetDefaultColumns returns a list of columns that are visible by default
+	GetDefaultColumns() []string
+
+	// SetColumnFilters sets additional column filters that will be used whenever one of the other methods of this
+	// interface are called. This is for example used to filter columns with information on kubernetes in a non-k8s
+	// environment like local-gadget
+	SetColumnFilters(...columns.ColumnFilter)
+
+	// SetSorting sets what sorting should be applied when calling SortEntries() // TODO
+	SetSorting([]string) error
+
+	// SetFilters sets which filter to apply before emitting events downstream
+	SetFilters([]string) error
+
+	// EventHandlerFunc returns a function that accepts an instance of type *T and pushes it downstream after applying
+	// enrichers and filters
+	EventHandlerFunc(enrichers ...func(any) error) any
+	EventHandlerFuncArray(enrichers ...func(any) error) any
+	EventHandlerFuncSnapshot(key string, enrichers ...func(any) error) any
+
+	// JSONHandlerFunc returns a function that accepts a JSON encoded event, unmarshal it into *T and pushes it
+	// downstream after applying enrichers and filters
+	JSONHandlerFunc(enrichers ...func(any) error) func([]byte)
+	JSONHandlerFuncArray(enrichers ...func(any) error) func([]byte)
+
+	// SetEventCallback sets the downstream callback
+	SetEventCallback(eventCallback any)
+
+	// SetErrorCallback sets the function to use whenever errors occur in this library or are received in-band
+	SetErrorCallback(errorCallback ErrorCallback)
+
+	// EnableSnapshots initializes the snapshot collector, which is able to aggregate snapshots from several sources
+	// and can return (optionally cached) results on demand; used for top gadgets
+	EnableSnapshots(ctx context.Context, t time.Duration, ttl int)
+}
+
+type parser[T any] struct {
+	columns            *columns.Columns[T]
+	sortBy             []string
+	filters            []string
+	filterSpec         *filter.FilterSpec[T] // TODO: filter collection(!)
+	eventCallback      func(*T)
+	eventCallbackArray func([]*T)
+	errorCallback      ErrorCallback
+	snapshotCombiner   *snapshotcombiner.SnapshotCombiner[T]
+	enableSnapshots    bool
+	columnFilters      []columns.ColumnFilter
+}
+
+func NewParser[T any](columns *columns.Columns[T]) Parser {
```

Comment:
Go question: What is the advantage of returning an interface here instead of directly returning the `parser` struct considering we don't expect to have multiple parsers (or do we?)?

Reply:
I think this is a good example for what you can do when you don't want to export internals. This way it's clear that you can store the result in a Parser variable/member (and you can actually have one of those vs unexported types). Returning an unexported/lowercase type would bring the issue that you would have to know about the interface type in the first place to be able to store the return value. In this case it also hides the type/generic nature of the actual implementation (otherwise you would also have to use generics whenever you would store the actual implementation directly).

Reply:
It is clear. Thanks!